{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "## Objectives\n",
        "We will be using `AlphaSimR` to  \n",
        "* Simulate datasets to test statistical genetic analysis methods  \n",
        "* Simulate whole breeding programs in view of optimizing them  \n",
        "\n",
        "Example code to work with `AlphaSimR` to  \n",
        "1. Create a population and get breeding values from `AlphaSimR` and from an\n",
        "\"experiment\"  \n",
        "2. Prepare a data structure for breeding records that will be useful for\n",
        "simulation.  \n",
        "3. Use this data structure in simple linear model analyses."
      ],
      "metadata": {
        "id": "A7To1OxY-0qO"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Script setup  \n",
        "Install packages"
      ],
      "metadata": {
        "id": "AX7ZSJ71_Q6v"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Loading libraries\n",
        "req_packages<-c(\"AlphaSimR\", \"tidyverse\")\n",
        "\n",
        "for(i in c(1:length(req_packages))){\n",
        "  if (!require(req_packages[i], character.only = TRUE)){\n",
        "   install.packages(req_packages[i])\n",
        "  }\n",
        "}"
      ],
      "metadata": {
        "id": "kXa5tC6M_Paq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Script parameters  "
      ],
      "metadata": {
        "id": "TQnQyciIAmTO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "random_seed <- 45678\n",
        "set.seed(random_seed)\n",
        "nFounders <- 100\n",
        "nChr <- 7\n",
        "segSites <- 140\n",
        "nQTL <- 100"
      ],
      "metadata": {
        "id": "WBccOKaeAqeP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Genotypic value versus breeding value  \n",
        "By default, the additive genetic variance among founders will be 1.\n",
        "Here are a couple of `AlphaSimR` genetic models with non-additive gene action. `addTraitAD` gives an additive + dominance gene action. Each locus has its own dominance deviation and there is a variance of it among loci:  \n",
        "$dd \\sim N(meanDD, varDD)$  \n",
        "`addTraitADE` adds pairwise additive by additive epistasis.  "
      ],
      "metadata": {
        "id": "-5Ec7cY1A4CA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Create haplotypes for founder population of outbred individuals\n",
        "founderHaps <- AlphaSimR::runMacs(nInd=nFounders, nChr=nChr, segSites=segSites)\n",
        "# Setup the genotype to phenotype mapping\n",
        "SP <- AlphaSimR::SimParam$new(founderHaps)\n",
        "meanDD <- 0.4\n",
        "varDD <- 0.04\n",
        "\n",
        "# Additive and dominance model\n",
        "SP$addTraitAD(nQtlPerChr=nQTL, meanDD=meanDD, varDD=varDD)\n",
        "\n",
        "# Create a new population of founders\n",
        "founders <- AlphaSimR::newPop(founderHaps, simParam=SP)\n",
        "\n",
        "# Compare genotypic and breeding values\n",
        "corBVGV <- cor(AlphaSimR::gv(founders), AlphaSimR::bv(founders)) %>% round(3)\n",
        "plot(AlphaSimR::gv(founders), AlphaSimR::bv(founders), pch=16,\n",
        "     xlab=\"Genotypic value\", ylab=\"Breeding value\",\n",
        "     main=paste0(\"BV vs GV: AD model, cor=\", corBVGV),\n",
        "     cex.axis=1.3, cex.lab=1.3)\n",
        "\n",
        "# Additive dominance and epistasis model\n",
        "# \"the relative value of additive-by-additive variance\n",
        "# compared to additive variance\"\n",
        "SP <- AlphaSimR::SimParam$new(founderHaps)\n",
        "relAA <- 0.5\n",
        "SP$addTraitADE(nQtlPerChr=nQTL, meanDD=meanDD, varDD=varDD, relAA=relAA)\n",
        "\n",
        "# Create a new population of founders\n",
        "founders <- AlphaSimR::newPop(founderHaps, simParam=SP)\n",
        "\n",
        "# Compare genotypic and breeding values\n",
        "corBVGV <- cor(AlphaSimR::gv(founders), AlphaSimR::bv(founders)) %>% round(3)\n",
        "plot(AlphaSimR::gv(founders), AlphaSimR::bv(founders), pch=16,\n",
        "     xlab=\"Genotypic value\", ylab=\"Breeding value\",\n",
        "     main=paste0(\"BV vs GV: ADE model, cor=\", corBVGV),\n",
        "     cex.axis=1.3, cex.lab=1.3)\n",
        ""
      ],
      "metadata": {
        "id": "Sm0fQqNZA8oS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Estimated versus analytical breeding value  \n",
        "Estimating the breeding value by a progeny test.  \n",
        "In practice, the BV can't be observed.  A progeny test calculates the average\n",
        "deviation from the population mean of progeny from randomly mating an individual\n",
        "to the population: 2 * mean(progeny pheno - pop mean)"
      ],
      "metadata": {
        "id": "qv27WqNgBHyr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Error variance for phenotypic evaluations\n",
        "varE <- 1\n",
        "# Number of progeny for breeding value estimation\n",
        "nProgeny1 <- 5\n",
        "nProgeny2 <- 50\n",
        "\n",
        "# Estimate breeding values\n",
        "# ind is the individual whose breeding value you want to estimate\n",
        "# pop is the population that individual is in\n",
        "# nProgeny is the number of progeny for the test\n",
        "# varE is the error variance with which phenotypes are evaluated\n",
        "#      if the genotypic variance is 1 then varE=1 will give h2 = 0.5\n",
        "estimateBV <- function(ind, pop, nProgeny, varE=1){\n",
        "  # I'm going to cheat a little and assume we know the population mean exactly\n",
        "  popMean <- AlphaSimR::gv(pop) %>% mean\n",
        "  # Set up crossPlan to cross ind to random others nProgeny times\n",
        "  crossPlan <- cbind(ind, sample(AlphaSimR::nInd(pop), nProgeny, replace=T))\n",
        "  progeny <- AlphaSimR::makeCross(pop, crossPlan)\n",
        "  progPheno <- AlphaSimR::setPheno(progeny, varE=varE, onlyPheno=T)\n",
        "  return(2*(mean(progPheno) - popMean))\n",
        "}\n",
        "\n",
        "# estimate BV with a progeny test of nProgeny1\n",
        "estimatedBV <- sapply(1:AlphaSimR::nInd(founders), estimateBV,\n",
        "                      pop=founders, nProgeny=nProgeny1, varE=varE)\n",
        "# Compare estimated and analytical breeding values\n",
        "plot(estimatedBV, AlphaSimR::bv(founders), pch=16,\n",
        "     xlab=\"Estimated value\", ylab=\"Analytical value\",\n",
        "     main=paste(\"Breeding value estimated from\", nProgeny1, \"Progeny\"))\n",
        "\n",
        "# estimate BV with a progeny test of nProgeny2\n",
        "estimatedBV <- sapply(1:AlphaSimR::nInd(founders), estimateBV,\n",
        "                      pop=founders, nProgeny=nProgeny2, varE=varE)\n",
        "# Compare estimated and analytical breeding values\n",
        "plot(estimatedBV, AlphaSimR::bv(founders), pch=16,\n",
        "     xlab=\"Estimated value\", ylab=\"Analytical value\",\n",
        "     main=paste(\"Breeding value estimated from\", nProgeny2, \"Progeny\"))"
      ],
      "metadata": {
        "id": "v_lDjk16BunE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Round robin crossing design  \n",
        "This is a simple design that ensures that all parents get used equally  "
      ],
      "metadata": {
        "id": "dKMMY_dSB0Af"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "makeRoundRobin <- function(pop, makeRandom=F){\n",
        "  nInd <- AlphaSimR::nInd(pop)\n",
        "  parOrder <- 1:nInd\n",
        "  if (makeRandom) parOrder <- sample(parOrder)\n",
        "  return(cbind(parOrder, parOrder[c(2:nInd, 1)]))\n",
        "}\n",
        "\n",
        "# Make a bunch of new lines\n",
        "crossPlan <- makeRoundRobin(founders)\n",
        "exptLines <- AlphaSimR::makeCross(founders, crossPlan)\n",
        "\n",
        "cbind(exptLines@id, exptLines@mother, exptLines@father) %>% head"
      ],
      "metadata": {
        "id": "iHDKAZuOB4xx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## `records` data structure  \n",
        "`AlphaSimR` populations only retain phenotypes of the most recent evaluation.\n",
        "In plant breeding, it is common to evaluate the same line more than once,\n",
        "and it makes sense to include all of those phenotypes in downstream analyses.\n",
        "Here, I propose a simple tibble to retain phenotypic records.  \n",
        "\n",
        "### Set up records\n",
        "The columns are the individual `id`, its seed parent id `seedPar`, its pollen\n",
        "parent id `pollenPar`, its observed phenotype `pheno`, and the error variance\n",
        "of that observed phenotype `varE`  "
      ],
      "metadata": {
        "id": "caqXjig0B-il"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "nInd <- 100\n",
        "varE <- 1\n",
        "# Function to make a simple data structure out of a population\n",
        "# AlphaSimR doesn't retain varE once you have setPheno, so supply it\n",
        "makeRecFromPop <- function(pop, varE=1){\n",
        "  dplyr::tibble(id=pop@id,\n",
        "                seedPar=pop@mother,\n",
        "                pollenPar=pop@father,\n",
        "                pheno=AlphaSimR::pheno(pop),\n",
        "                varE=varE)\n",
        "}\n",
        "\n",
        "# Run a phenotyping experiment and populate the `records` data structure\n",
        "exptLines <- AlphaSimR::setPheno(exptLines, varE=varE, simParam=SP)\n",
        "records <- makeRecFromPop(exptLines, varE=varE)\n",
        "head(records)\n",
        "\n",
        "# Compare genotypic and phenotypic values\n",
        "corPhGV <- cor(AlphaSimR::gv(exptLines), AlphaSimR::pheno(exptLines)) %>%\n",
        "  round(3)\n",
        "plot(AlphaSimR::gv(exptLines), AlphaSimR::pheno(exptLines), pch=16,\n",
        "     xlab=\"Genetic value\", ylab=\"Phenotype\",\n",
        "     main=paste0(\"First Evaluation, cor=\", corPhGV))\n",
        "\n",
        "# Second evaluation\n",
        "exptLines <- AlphaSimR::setPheno(exptLines, varE=varE, simParam=SP)\n",
        "# Add the new records to the old ones\n",
        "records <- dplyr::bind_rows(records, makeRecFromPop(exptLines, varE=varE))\n",
        "\n",
        "# Compare genotypic and phenotypic values\n",
        "corPhGV <- cor(AlphaSimR::gv(exptLines), AlphaSimR::pheno(exptLines)) %>%\n",
        "  round(3)\n",
        "plot(AlphaSimR::gv(exptLines), AlphaSimR::pheno(exptLines), pch=16,\n",
        "     xlab=\"Genetic value\", ylab=\"Phenotype\",\n",
        "     main=paste0(\"Second Evaluation, cor=\", corPhGV))"
      ],
      "metadata": {
        "id": "d6G25OIhCGOq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Use records to estimate genotypic value from >1 observation"
      ],
      "metadata": {
        "id": "wx0HF91JCMH6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# BASIC LINEAR MODEL\n",
        "lmEstGV <- lm(pheno ~ -1 + id, data=records)\n",
        "gvEstimates <- coefficients(lmEstGV)\n",
        "\n",
        "# Compare genotypic and **Combined** phenotypic values\n",
        "corEstGV <- cor(AlphaSimR::gv(exptLines), gvEstimates) %>%\n",
        "  round(3)\n",
        "plot(AlphaSimR::gv(exptLines), gvEstimates, pch=16,\n",
        "     xlab=\"Genetic value\", ylab=\"Phenotype\",\n",
        "     main=paste0(\"Both Evaluations, cor=\", corEstGV))"
      ],
      "metadata": {
        "id": "Xh6PMuSDCRm-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Use records in two-stage selection\n",
        "In plant breeding there is often a preliminary evaluation stage, then a\n",
        "more advanced stage.  We can use the records to keep track of these stages."
      ],
      "metadata": {
        "id": "Tk-m2xCnCalk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "nIndStage1 <- 400\n",
        "nIndStage2 <- 100\n",
        "varEstage1 <- 4\n",
        "varEstage2 <- 1\n",
        "\n",
        "progenyPerCross <- nIndStage1 / nrow(crossPlan)\n",
        "# The production pipeline starts with a bunch of new lines\n",
        "exptLines <- AlphaSimR::makeCross(founders, crossPlan,\n",
        "                                  nProgeny=progenyPerCross)\n",
        "\n",
        "# Phenotypic evaluation of experimental lines\n",
        "exptLines <- AlphaSimR::setPheno(exptLines, varE=varEstage1, simParam=SP)\n",
        "records <- makeRecFromPop(exptLines, varE=varEstage1)\n",
        "\n",
        "# Select among lines to advance to Stage 2\n",
        "keep <- AlphaSimR::pheno(exptLines) %>%\n",
        "  order(decreasing=T) %>% .[1:nIndStage2]\n",
        "\n",
        "# Phenotypic evaluation of Stage 2 lines\n",
        "stage2Lines <- exptLines[keep]\n",
        "stage2Lines <- AlphaSimR::setPheno(stage2Lines, varE=varEstage2, simParam=SP)\n",
        "records <- dplyr::bind_rows(records, makeRecFromPop(stage2Lines,\n",
        "                                                    varE=varEstage2))\n",
        "\n",
        "str(records)"
      ],
      "metadata": {
        "id": "ZKpo7jubCf8_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Estimate gain from selection four ways\n",
        "This little section is mostly to illustrate that `AlphaSimR` thinks of\n",
        "breeding values *relative to the population mean*, but of genotypic values in\n",
        "*absolute* terms."
      ],
      "metadata": {
        "id": "rpadSZPlCmrh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"Gain from selection on BV where the reference pop is stage2Lines\")\n",
        "print(paste(\"Gain from selection\",\n",
        "            round(mean(AlphaSimR::bv(stage2Lines)) -\n",
        "                  mean(AlphaSimR::bv(exptLines)), 2)))\n",
        "\n",
        "print(\"Gain from selection on BV where the reference pop is all exptLines\")\n",
        "print(paste(\"Gain from selection\",\n",
        "            round(mean(AlphaSimR::bv(exptLines)[keep]) -\n",
        "                  mean(AlphaSimR::bv(exptLines)), 2)))\n",
        "\n",
        "print(\"Gain from selection on **GV** where the reference pop is stage2Lines\")\n",
        "print(paste(\"Gain from selection\",\n",
        "            round(mean(AlphaSimR::gv(stage2Lines)) -\n",
        "                  mean(AlphaSimR::gv(exptLines)), 2)))\n",
        "\n",
        "print(\"Gain from selection on **GV** where the reference pop is all exptLines\")\n",
        "print(paste(\"Gain from selection\",\n",
        "            round(mean(AlphaSimR::gv(exptLines)[keep]) -\n",
        "                  mean(AlphaSimR::gv(exptLines)), 2)))"
      ],
      "metadata": {
        "id": "9okgODMFC5x3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Homework  \n",
        "1. (**3 pts**) We saw how to write a function to estimate breeding values by doing a progeny\n",
        "test (chunk `Estimate breeding value`). Then, in the chunk `Linear model\n",
        "estimation`, we used a simple linear model to combine records to estimate the\n",
        "genotypic values of individuals (chunk `Linear model estimation`).  \n",
        "Tweak the linear model so that it estimates, for example, the breeding\n",
        "values of the seed parent. HINT: this will entail introducing `seedPar`\n",
        "into the model. Use the `exptLines` and their associated `records` from the\n",
        "`Two stage selection` chunk.  \n",
        "2. (**2 pts**) Make plots comparing the breeding values thus obtained to those\n",
        "obtained by the `bv` function or by our own `estimateBV` function.  \n",
        "3. (**3 pts**) The `lm` function from the `stats` package can account for observations with\n",
        "different weights. Weights should be the inverse of the error variance for the\n",
        "observation. Handily, we keep that error variance in the `records` tibble.\n",
        "Again, tweak the linear model and use the records from the `Two stage selection`\n",
        "chunk to account for the different error variances from Stage 1 versus Stage 2\n",
        "when you estimate the genotypic values for individuals.  \n",
        "4. (**2 pts**) Correlate the genotypic values obtained with or without weighting to the\n",
        "simulated values obtained by the `AlphaSimR::gv` function.  "
      ],
      "metadata": {
        "id": "1e1OsO4AC7C4"
      }
    }
  ]
}