{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "### Learning objectives\n",
        "\n",
        "1. AlphaSimR  \n",
        "`AlphaSimR` is a package to simulate breeding populations and tasks. It is not\n",
        "completely intuitive nor completely well-documented. We will want to use it\n",
        "ultimately to optimize breeding schemes. As a way to introduce AlphaSimR, this\n",
        "Rmarkdown script contains code to illustrate the \"outbreak of variation\" that\n",
        "occurs when a heterozygous individual is self-fertilized. This phenomenon was\n",
        "important in supporting the notion that quantitative traits were influenced by\n",
        "many loci.  Here is an\n",
        "[introduction](https://cran.r-project.org/web/packages/AlphaSimR/vignettes/intro.html)\n",
        "showing code to simulate a very simple mass selection program.  Here is a\n",
        "[deep dive](https://cran.r-project.org/web/packages/AlphaSimR/vignettes/traits.pdf)\n",
        "into how `AlphaSimR` simulates traits.\n",
        "\n",
        "2. Prepare a Homework  \n",
        "Your homework will be to write a script that illustrates the interesting phenomenon of regression toward the mean between parents and progeny that we discussed in class.  \n",
        "\n",
        "# Ordering of the script  \n",
        "It's good to have all scripts in the same order with respect to standard tasks  \n",
        "\n",
        "### Load packages first  \n",
        "If your script depends on external packages, load them at the beginning. This shows users early on what the script dependencies are.  "
      ],
      "metadata": {
        "id": "CD1jMufW6j3D"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Loading libraries\n",
        "req_packages<-c(\"AlphaSimR\", \"tidyverse\")\n",
        "\n",
        "for(i in c(1:length(req_packages))){\n",
        "  if (!require(req_packages[i], character.only = TRUE)){\n",
        "   install.packages(req_packages[i])\n",
        "  }\n",
        "}"
      ],
      "metadata": {
        "id": "k7MUnD2t7ZVC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Notice the conflicts report from loading tidyverse.  Two packages (`dplyr` and `stats`) both have a function called `filter`. Since dplyr was loaded *after* stats, if you use the function `filter`, it will go to the dplyr version.  It is dangerous to rely on what order packages have been loaded to determine which `filter` function you get.  R syntax to prevent ambiguity is to write either `dplyr::filter` or `stats::filter`.  Using that syntax will make your code more reproducible.\n",
        "\n",
        "### Set random seed  \n",
        "AlphaSimR generates many random numbers (e.g., to simulate Mendelian random segregation).  If you want the result of the analysis to come out the same each time (there are pros and cons) you need to set the random seed. Note that `workflowr` does this by default. If you are not using that package, then be explicit."
      ],
      "metadata": {
        "id": "vD9laLdqmbwm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "random_seed <- 45678\n",
        "set.seed(random_seed)"
      ],
      "metadata": {
        "id": "bMHGe0gxmuul"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Script parameters  \n",
        "If the behavior of your script depends on parameters that you set, initialize them early on."
      ],
      "metadata": {
        "id": "KktzYGEUmzED"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "nFounders <- 100\n",
        "nChr <- 10 # Number of chromosomes\n",
        "nSitesPerChr <- 1000 # Number of segregating sites _per chromosome_\n",
        "nQTLperChr <- 10 # Vary this parameter to get oligo- versus poly- genic traits\n",
        "nF1s <- 200 # We are going to make F1s to test outbreak of variation\n",
        "nF2s <- 200 # We are going to make F2s to test outbreak of variation"
      ],
      "metadata": {
        "id": "Wvih8Ttgm5Ka"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Simulating some classical results  \n",
        "This script uses `AlphaSimR` to simulate the \"outbreak of variation\" that arises when you self-fertilize a hybrid.  \n",
        "\n",
        "### AlphaSimR populations  \n",
        "The basic object of `AlphaSimR` is the population. To make founders, you first make founder haplotypes from a coalescent simulation, then you define simulation parameters that will link their genetic variation to phenotypic variation, then you make a first diploid population from the founder haplotypes.  "
      ],
      "metadata": {
        "id": "psXLkpOP6UuG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Create haplotypes for founder population of outbred individuals\n",
        "# Note: default effective population size for runMacs is 100\n",
        "founderHaps <- AlphaSimR::runMacs(nInd=nFounders, nChr=nChr,\n",
        "                                  segSites=nSitesPerChr)\n",
        "# founderHaps <- AlphaSimR::quickHaplo(nInd=nFounders, nChr=nChr,\n",
        "#                                  segSites=nSitesPerChr)\n",
        "\n",
        "# New global simulation parameters from founder haplotypes\n",
        "SP <- AlphaSimR::SimParam$new(founderHaps)\n",
        "# Additive trait architecture\n",
        "# By default, the genetic variance will be 1\n",
        "SP$addTraitA(nQtlPerChr=nQTLperChr)\n",
        "\n",
        "# Create a new population of founders\n",
        "founders <- AlphaSimR::newPop(founderHaps, simParam=SP)\n",
        "str(founders)"
      ],
      "metadata": {
        "id": "1zlLfIO96bCp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Population information  \n",
        "The population has `id`s. The `@mother` and `@father` ids are all zero because this population was made from founder haplotypes, and so does not have diploid parents. The genotypic values `gv` of the population are calculated for the trait created using `SP$addTraitA(nQtlPerChr=nQTLperChr)`. Given that there are `r nChr` chromosome`r ifelse(nChr > 1, \"s\", \"\")` and `r nQTLperChr` QTL per chromosome, there are `3^(nChr*nQTLperChr)` = `r 3^(nChr*nQTLperChr)` different possible genotypic values. The realized genotypic values are accessible with the function `gv(founders)`\n",
        "\n",
        "From here, you can treat this population like a named vector using the square braces extraction operator `[ ]`.  Extract individuals by their `@id` or just by their order in the population using an integer index. For example, pick three random individuals from a population and list their ids. Pick the one with the first id in alphabetical order."
      ],
      "metadata": {
        "id": "7M-p2BfA6k_J"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "test <- founders[c(2, 3, 5, 8, 13)]\n",
        "testID <- test@id\n",
        "firstID <- testID[order(testID)][1]\n",
        "alphaInd <- test[firstID] # Extract individual with the first ID in order\n",
        "print(testID)\n",
        "print(alphaInd)\n",
        "print(alphaInd@id)"
      ],
      "metadata": {
        "id": "dpYxhzPU6qZ0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Outbreak of variation  \n",
        "Emerson and East (1913) showed that if you crossed two inbreds, the hybrid had similar variation to each inbred, but if you then selfed the hybrid, the offspring varied more substantially. This code simulates that result. First, self the founders to homozygosity. The function `self` self-fertilizes individuals from the population. By default, it creates one selfed individual per parent (controllable with the parameter `nProgeny`), so this works nicely for single-seed descent.  "
      ],
      "metadata": {
        "id": "joOSBIk46wGY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Self-pollinate to for a few generations\n",
        "nGenSelf <- 3\n",
        "inbredPop <- founders\n",
        "for (gen in 1:nGenSelf){\n",
        "  inbredPop <- AlphaSimR::self(inbredPop)\n",
        "}"
      ],
      "metadata": {
        "id": "tsLZiQnY6zh3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Check homozygosity  \n",
        "Just a sanity check that this has, in fact, created a population of 100 individuals that are appropriately homozygous.  Loci are coded 0, 1, 2. So `qtl == 1` represents the case were a locus is heterozygous. `sum(qtl == 1)` counts those cases."
      ],
      "metadata": {
        "id": "jyE71Pw57EWO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "qtl <- AlphaSimR::pullQtlGeno(inbredPop)\n",
        "if (nrow(qtl) != nFounders) stop(\"The number of individuals is unexpected\")\n",
        "if (ncol(qtl) != nChr * nQTLperChr) stop(\"The number of QTL is unexpected\")\n",
        "fracHet <- sum(qtl == 1) / (nFounders * nChr * nQTLperChr)\n",
        "cat(\"Expected fraction heterozygous\", 1 / 2^nGenSelf, \"\\n\",\n",
        "    \"Observed fraction heterozygous\", fracHet, \"+/-\",\n",
        "    round(2*sqrt(fracHet*(1-fracHet)/(nFounders*nChr*nQTLperChr)), 3), \"\\n\")"
      ],
      "metadata": {
        "id": "dVS1I2R77IP-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "> What was wrong with my reasoning about the Expected fraction heterozygous?\n",
        "\n",
        "### Simulate outbreak of variation  \n",
        "We will  \n",
        "1. pick a random pair of inbred individuals  \n",
        "2. cross that pair  \n",
        "3. find out the variation in genotypic value among the pair's progeny  \n",
        "4. pick a random F1 progeny  \n",
        "5. self-fertilize that F1  \n",
        "6. find out the variation in genotypic value among the F1's progeny  \n",
        "*It's often useful to write very high-level outlines of what you want your code to do before you write the code*  \n",
        "We will assume a trait that has a heritability of 0.5 in the base, non-inbred population. An easy way to get that heritability is with genetic and error variances of 1.\n",
        "\n",
        "### Pick random pair to make F1 hybrid"
      ],
      "metadata": {
        "id": "x6ybNhIX7o17"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "randomPair <- inbredPop[sample(nFounders, 2)]"
      ],
      "metadata": {
        "id": "V-V88PKC8gFI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Cross the pair to make population of F1s  \n",
        "The crossPlan is a **two-column matrix** with as many rows as the number of crosses you want to make, the first column is the `@id` or the index of the seed parent, and likewise for the pollen parent in the second column.  We will make `r nF1s` F1s from this random pair, so the matrix has `r nF1s` rows. You just want to cross individual 1 with individual 2, so each row contains 1:2."
      ],
      "metadata": {
        "id": "Jdtu-YIX8rtG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "crossPlan <- matrix(rep(1:2, nF1s), ncol=2, byrow=T)\n",
        "f1_pop <- AlphaSimR::makeCross(randomPair, crossPlan)\n",
        "head(crossPlan)"
      ],
      "metadata": {
        "id": "0yqUB9Tc8yxN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Find out the phenotypic variation among F1s  \n",
        "When you first make a population, `AlphaSimR` does not assume it has been phenotyped. You can phenotype it using the `setPheno` function.  Note that if you use `setPheno` on the same population a second time, that will overwrite the phenotypes from the first time. The genotypic variance can be retrieved using the function `varG`. Really, `varG` gives all we need to know, but of course that variance is generally not observable in non-simulated reality."
      ],
      "metadata": {
        "id": "n5GyL9wG9Qjy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "f1_pop <- AlphaSimR::setPheno(f1_pop, varE=1)\n",
        "cat(\"Genotypic variance among F1s\", round(AlphaSimR::varG(f1_pop), 3), \"\\n\")\n",
        "cat(\"Phenotypic variance among F1s\", round(AlphaSimR::varP(f1_pop), 3), \"\\n\")\n",
        "hist(AlphaSimR::pheno(f1_pop), main=\"Histogram of F1 Phenotypes\")"
      ],
      "metadata": {
        "id": "iO_Vlky19YRV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Pick random F1"
      ],
      "metadata": {
        "id": "MwpDRslj_4nb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "randomF1 <- f1_pop[sample(nFounders, 1)]"
      ],
      "metadata": {
        "id": "d9ZS8VhX_97E"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Make F2 and observe phenotypic variance  \n",
        "When you first make a population, `AlphaSimR` does not assume it has been phenotyped. You can phenotype it using the `setPheno` function.  Note that if you use `setPheno` on the same population a second time, that will overwrite the phenotypes from the first time. The genotypic variance can be retrieved using the function `varG`. Really, `varG` gives all we need to know, but of course that variance is generally not observable in non-simulated reality."
      ],
      "metadata": {
        "id": "yEkfrMmf_b71"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "f2_pop <- AlphaSimR::self(randomF1, nProgeny=nF2s)\n",
        "f2_pop <- AlphaSimR::setPheno(f2_pop, varE=1)\n",
        "cat(\"Genotypic variance among F2s\", round(AlphaSimR::varG(f2_pop), 3), \"\\n\")\n",
        "cat(\"Phenotypic variance among F2s\", round(AlphaSimR::varP(f2_pop), 3), \"\\n\")\n",
        "hist(AlphaSimR::pheno(f2_pop), main=\"Histogram of F2 Phenotypes\")"
      ],
      "metadata": {
        "id": "AByC6wJs_tqn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Homework  \n",
        "Illustrate regression to the mean between parents and offspring using AlphaSimR  \n",
        "1. (**1 pt**) You know how to create a population -- create a population of parents\n",
        "2. (**1 pt**) You know how to get phenotypes from that population  \n",
        "3. (**2 pts**) You want to randomly mate that population to get progeny that will be (we assume)\n",
        "regressed to the mean. Create a `crossPlan` matrix like I did to generate the\n",
        "F1s, except that each row should have randomly-picked parents, rather than 1 and\n",
        "2 like for the F1s. There is also a command `AlphaSimR::randCross`.  Check out\n",
        "its documentation. If you use that command, you will have to find the pedigree\n",
        "of the progeny using the `@mother` and `@father` ids of the progeny population\n",
        "and match those up to the parent population. If you make the crossPlan, then it\n",
        "gives you the seed and pollen parent ids.\n",
        "4. (**1 pt**) Having made the progeny population, phenotype it also  \n",
        "5. (**2 pts**) Use each row of the `crossPlan` to find the two parents and calculate their phenotypic mean  \n",
        "6. (**1 pts**) Make a scatterplot of the progeny phenotypes against the parent mean phenotypes  \n",
        "7. (**2 pts**) What is the regression coefficient? Was there regression towards the mean?\n",
        "\n",
        "### Homework grading  \n",
        "  \n",
        "I will subtract 0.5 points for every day the homework is late  "
      ],
      "metadata": {
        "id": "gbK53i9U9wUY"
      }
    }
  ]
}